### PR TITLE
Overhaul SDK Metrics

### PR DESCRIPTION
**Pull request by dianaafanador3**
Overhauled SDK metrics collection to better drive future development effort.
After the PR we still send the metrics the first time a realm is opened in the life time of and App on Debug, but we now do it after we have the schema, so we can get some features metrics.

Here is a list of the metrics, and info about if is tracked or not.
Field Name | isTracked
-- | --
distinct_id | ✅
builder_id | ✅
Anonymized Bundle ID | ✅
Binding | ✅
Language | ✅
Host OS Type | ✅
Host OS Version | ✅
Host CPU Arch | ✅
Target CPU Arch | ✅
Target OS Type | ✅
Target OS Minimum Version | ✅
Target OS Version | ✅
Realm Version | ✅
Core Version | ✅
Sync Enabled | ✅
Language Version |  
Framework | ✅
Framework Version |  
Installation Method | ✅
Installation Method Version |  
IDE |  
IDE Version | 
Runtime Engine |  
Compiler | ✅
Embedded_Object | ✅
Asymmetric_Object | ✅
Reference_Link | ✅
Mixed | ✅
Primitive_List | ✅
Primitive_Set | ✅
Primitive_Dictionary | ✅
Reference_List | ✅
Reference_Set | ✅
Reference_Dictionary | ✅
Backlink | ✅
Realm_Integer |  
Asynchronous_Realm_Open |  
Synchronous_Realm_Open |  
Query_Async |  
Query_Primary_Key |  
Write_Async |  
Thread_Safe_Reference |  
Insert_Modified |  
Compact_On_Launch | ✅
Schema_Migration_Block | ✅
Realm_Change_Listener |  
List_Change_Listener |  
Set_Change_Listener |  
Dictionary_Change_Listener |  
Result_Change_Listener |  
Object_Change_Listener |  
Client_Reset_Recover_Or_Discard | ✅
Client_Reset_Recover | ✅
Client_Reset_Discard | ✅
Client_Reset_Manual | ✅
Progress_Notification |  
Connection_Notification |  
Pbs_Sync | ✅
Flexible_Sync | ✅
Auth_Anonymous |  
Auth_Email_Password |  
Auth_Facebook |  
Auth_Google |  
Auth_Custom_JWT |  
Auth_API_Key |  
Auth_Server_API_Key |  
Auth_Apple |  
Auth_Function |  
Remote_Function |  
MongoDB_Data_Access |  
Dynamic_API |  
String_FTS |

### CODE REVIEW

File: Realm/RLMAnalytics.mm [hunk 1]
```diff
@@ -194,73 +296,206 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+        @"Mixed": @0,
+        @"Primitive_List": @0,
+        @"Primitive_Set": @0,
+        @"Primitive_Dictionary": @0,
+        @"Reference_List": @0,
+        @"Reference_Set": @0,
+        @"Reference_Dictionary": @0,
+    };
+    NSMutableDictionary *featuresDictionary = [[NSMutableDictionary alloc] init];
+    [featuresDictionary addEntriesFromDictionary: dict];
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:401]:
```suggestion
    NSMutableDictionary *featuresDictionary = [@{
        @"Embedded_Object": @0,
        @"Asymmetric_Object": @0,
        @"Reference_Link": @0,
        @"Mixed": @0,
        @"Primitive_List": @0,
        @"Primitive_Set": @0,
        @"Primitive_Dictionary": @0,
        @"Reference_List": @0,
        @"Reference_Set": @0,
        @"Reference_Dictionary": @0,
    } mutableCopy];
```

File: Realm/RLMAnalytics.mm [hunk 2]
```diff
@@ -194,73 +296,206 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+    for (RLMObjectSchema *objectSchema in schema.objectSchema) {
+        if (objectSchema.isEmbedded) {
+            [featuresDictionary setObject:@1 forKey:@"Embedded_Object"];
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:404]:
```suggestion
            featuresDictionary[@"Embedded_Object"] = @1;
```

(this applies to all of these)

File: Realm/RLMAnalytics.mm [hunk 3]
```diff
@@ -194,73 +296,206 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+               default:
+                    break;
+            }
+        }
+    }
+    return featuresDictionary;
+}
+
+static NSDictionary *RLMConfigurationMetrics(RLMRealmConfiguration *configuration) {
+    BOOL isSyncEnable = configuration.syncConfiguration != nil;
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:455]: Every read of the syncConfiguration property is allocating a new object, which should be cached.

BOOL is a bad obj-c type (it's a typedef for `char` rather than `bool`) and should generally only appear at public API boundaries.

File: Realm/RLMAnalytics.mm [hunk 4]
```diff
@@ -194,73 +296,206 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+    NSDictionary *schemaMetrics = RLMSchemaMetrics(schema);
+    NSDictionary *configurationMetrics = RLMConfigurationMetrics(configuration);
+
+    NSMutableDictionary *metrics = [[NSMutableDictionary alloc] init];
+    [metrics addEntriesFromDictionary:baseMetrics];
+    [metrics addEntriesFromDictionary:schemaMetrics];
+    [metrics addEntriesFromDictionary:configurationMetrics];
+
+    NSMutableDictionary *payloadN = [NSMutableDictionary dictionaryWithDictionary:@{ @"event": @"Run" }];
+    [payloadN setObject:metrics forKey:@"properties"];
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:497]:
```suggestion
    NSDictionary *payloadN = @{@"event": @"Run", @"properties": metrics};
```

File: Realm/RLMAnalytics.mm [hunk 5]
```diff
@@ -194,73 +296,206 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+        @"Client_Reset_Discard": (isSync && resetMode == RLMClientResetModeDiscardUnsyncedChanges) ? @1 : @0,
+        @"Client_Reset_Manual": (isSync && resetMode == RLMClientResetModeManual) ? @1 : @0,
+
+        // Configuration
+        @"Compact_On_Launch": isCompactOnLaunch ? @1 : @0,
+        @"Schema_Migration_Block": migrationBlock ? @1 : @0,
+    };
+}
+
+void RLMSendAnalytics(RLMRealmConfiguration *configuration, RLMSchema *schema) {
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:481]: This is now doing quite a lot of work, so it should dispatch that work to a background queue rather than blocking opening the Realm while it prepares the analytics payload.

File: Realm/RLMAnalytics.mm [hunk 6]
```diff
@@ -194,73 +296,206 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+    BOOL isPBSSync = configuration.syncConfiguration.partitionValue != nil;
+    BOOL isFlexibleSync = (isSync && !isPBSSync);
+    BOOL isCompactOnLaunch = configuration.shouldCompactOnLaunch != nil;
+    BOOL migrationBlock = configuration.migrationBlock != nil;
+
+    auto resetMode = configuration.syncConfiguration.clientResetMode;
+    return @{
+        // Sync
+        @"Sync Enabled": isSyncEnable ? @"true" : @"false",
+        @"Flexible_Sync": isFlexibleSync ? @1 : @0,
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:466]: All these places where we're reporting 1/0 instead of "true"/"false" can just be `@(isFlexibleSync)`.

File: Realm/RLMAnalytics.mm [hunk 7]
```diff
@@ -101,22 +105,99 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+    sysctlbyname("kern.osproductversion", NULL, &size, NULL, 0);
+    char *model = (char*)malloc(size);
+    sysctlbyname("kern.osproductversion", model, &size, NULL, 0);
+    NSString *deviceModel = [NSString stringWithCString:model encoding:NSUTF8StringEncoding];
+    free(model);
+    return deviceModel;
+}
+
+static NSString *RLMHostArch() {
+    std::array<int, 2> mib = {{CTL_HW, HW_MACHINE}};
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:119]: This does not get the host architecture, but we also don't report analytics in any scenario where the host and target architectures would be different so it doesn't really make sense to do in the first place.

File: Package.swift [hunk 8]
```diff
@@ -97,6 +98,37 @@ func objectServerTestTarget(name: String, sources: [String]) -> Target {
     )
 }
 
+func runCommand() -> String {
```
Comment (tgoyne) [inline Package.swift:101]: This won't work when actually installing the package as a dependency. Dependency packages are run in a sandbox which doesn't have access to anything (not even the package's own files, which is why we can't read the version from dependencies.list). `swift build` doesn't sandbox the local package being built so it'll work there but not in any other scenario.

Comment (dianaafanador3) [reply Package.swift:101]: @tgoyne do you have a tip how we can get this IOPlatformUUID? Or should we not use this?

File: Realm/RLMAnalytics.mm (hunk 6 above)
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm:466]: This is a requirement on the scope document, all the features should report 1 or 0 if true or not.

Comment (tgoyne) [reply Realm/RLMAnalytics.mm:466]: Yes, and that's what `@(isFlexibleSync)` will do.

File: Package.swift (hunk 8 above)
Comment (tgoyne) [reply Package.swift:101]: I don't think there's any way to get it that works for any installation methods other than cocoapods.

Comment (dianaafanador3) [reply Package.swift:101]: Tested this on Cocoapods and SPM and it is working, tested this on debug and also taking a look at the logs in our metrics app on Mongo. Checking now if works as well in Carthage and XCFramework, I don't know if I'm missing a another use where this will not work

File: Realm/RLMAnalytics.mm [hunk 9]
```diff
@@ -101,22 +105,99 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+    targetArchitecture = @"x86_64";
+#elif TARGET_CPU_ARM
+    targetArchitecture = @"arm";
+#elif TARGET_CPU_ARM64
+    targetArchitecture = @"arm64";
+#endif
+    return targetArchitecture;
+}
+
+static NSString *RLMXCodeVersion() {
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm:147]: Are we mapping this clang version to Xcode version in our analytics dashboard?, Because if it is something hard to do there (the team doing doesn't have the resources), it will be better to do it in our side

File: Realm/RLMAnalytics.mm [hunk 10]
```diff
@@ -101,22 +105,99 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
 
 // Get the version of OS X we're running on (even in the simulator this gives
 // the OS X version and not the simulated iOS version)
-static NSString *RLMOSVersion() {
-    std::array<int, 2> mib = {{CTL_KERN, KERN_OSRELEASE}};
+static NSString *RLMHostOSVersion() {
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm:108]: we are already reporting this as the Target version

[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMAnalytics.mm:108]: Then why are we reporting the same thing twice?

[the remark this answers is not in the document]
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm:108]: "Host OS Version" reports the version of the Mac they are using to run the project, in my case :"13.2.1" and "Target OS Version" reports the os version of the target, in my case running on simulator 16.1 Version 16.1 (Build 20B72)

File: Realm/RLMAnalytics.mm (hunk 9 above)
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm]: I removed it

File: scripts/generate-rlmplatform.sh [hunk 11]
```diff
@@ -12,6 +12,7 @@ if [ "$IS_MACCATALYST" = "YES" ]; then
 fi
 
 unifdef -B -DREALM_BUILDING_FOR_$(echo ${PLATFORM_SUFFIX} | tr "[:lower:]" "[:upper:]") < "${SOURCE_FILE}" | sed -e "s/''/'/" > "${TEMPORARY_FILE}"
+echo "#define REALM_IOPLATFORMUUID @\"$(sh build.sh get-ioplatformuuid)\"" >> ${TEMPORARY_FILE}
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply scripts/generate-rlmplatform.sh]: Removed this

File: Package.swift (hunk 8 above)
Comment (dianaafanador3) [reply Package.swift:101]: We are only using the IOPlatformUUID with SPM and Cocoapods, for other types of installation we will return the Mac address (same as `distinct_id`). This will be helpful in some cases (if a user is using SPM or Cocoapods and is using another SDK), but we don't want to be strict about this info.

**Comment by dianaafanador3** [comment]
Putting this on review @tgoyne, the only missing change is the endpoin.

File: Realm/RLMRealm.mm [hunk 12]
```diff
@@ -477,6 +478,9 @@ + (instancetype)realmWithConfiguration:(RLMRealmConfiguration *)configuration
     }
 #endif
 
+    // Run Analytics and Update checker, this will be run only the first any realm open
+    [self runFirstCheckForConfiguration:configuration schema:realm.schema];
```
Comment (tgoyne) [inline Realm/RLMRealm.mm:482]: This is moving the call to set_default_logger() to too late in the opening process.

File: Realm/RLMRealm.mm [hunk 13]
```diff
@@ -80,6 +80,7 @@ void RLMDisableSyncToDisk() {
 }
 
 static std::atomic<bool> s_set_skip_backup_attribute{true};
+static bool initialized;
```
Comment (tgoyne) [inline Realm/RLMRealm.mm:83]: I don't see a reason this variable's scope needs to be expanded?

File: Realm/RLMPlatform.h.in [hunk 14]
```diff
@@ -45,3 +45,5 @@
 #error Attempting to use Realm''s watchOS framework in a non-watchOS target.
 #endif
 #endif
+
+#define REALM_IOPLATFORMUUID @""
```
Comment (tgoyne) [inline Realm/RLMPlatform.h.in:49]: Rather than putting this in an unrelated file it should be set in the xcconfig.

File: Realm/RLMAnalytics.mm (hunk 7 above)
Comment (tgoyne) [reply Realm/RLMAnalytics.mm:119]: There used to be i386 simulators which would run on x86_64, but these days there's only x86_64 and arm64.

File: Realm/RLMAnalytics.mm [hunk 15]
```diff
@@ -194,80 +241,207 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+void RLMSendAnalytics(RLMRealmConfiguration *configuration, RLMSchema *schema) {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        if (getenv("REALM_DISABLE_ANALYTICS") || !RLMIsDebuggerAttached() || RLMIsRunningInPlayground()) {
+            return;
+        }
+        NSArray *urlStrings = @[@"https://data.mongodb-api.com/app/realmsdkmetrics-zmhtm/endpoint/metric_webhook/metric?data=%@"];
+
+        NSDictionary *baseMetrics = RLMBaseMetrics();
+        NSDictionary *schemaMetrics = RLMSchemaMetrics(schema);
+        NSDictionary *configurationMetrics = RLMConfigurationMetrics(configuration);
```
Comment (tgoyne) [inline Realm/RLMAnalytics.mm:428]: configuration needs to be copied before the dispatch_async() as it's a mutable object and the user could mutate after opening a Realm while we're reading it on a background thread.

File: Realm/RLMAnalytics.mm [hunk 16]
```diff
@@ -194,80 +241,207 @@ static auto RLMSysCtl(int *mib, u_int mibSize, size_t *bufferSize) {
...
+        @"Schema_Migration_Block": migrationBlock ? @1 : @0,
+    };
+}
+
+void RLMSendAnalytics(RLMRealmConfiguration *configuration, RLMSchema *schema) {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        if (getenv("REALM_DISABLE_ANALYTICS") || !RLMIsDebuggerAttached() || RLMIsRunningInPlayground()) {
+            return;
+        }
+        NSArray *urlStrings = @[@"https://data.mongodb-api.com/app/realmsdkmetrics-zmhtm/endpoint/metric_webhook/metric?data=%@"];
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm:424]: This was already implemented like this, I guess this was used when there was more than one endpoint, or during testing. I'm afraid that if we change this, we will have to revert it in the future.

File: Realm/RLMRealm.mm (hunk 12 above)
Comment (dianaafanador3) [reply Realm/RLMRealm.mm:482]: I moved to `initialize`, but this will be replaced by the implementation on the Core's Logging PR when we are actually setting a default logger and a level threshold.

File: Realm/RLMAnalytics.mm (hunk 7 above)
Comment (dianaafanador3) [reply Realm/RLMAnalytics.mm:119]: mmmm I guess you are right, there is no case where the host architecture is different that the target architecture, I guess we could inform at least one of them. @tgoyne

File: Realm/RLMAnalytics.mm (hunk 16 above)
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm/RLMAnalytics.mm:406]: We had two endpoints when we were sending analytics to both mixpanel and our internal platform, and this should have been cleaned up when we dropped mixpanel.

File: Realm/RLMRealm.mm (hunk 13 above)
Comment (tgoyne) [reply Realm/RLMRealm.mm:83]: This appears to now just be unused.

**Review by tgoyne — APPROVED** [review]
